frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSquashAction.java

```diff
@@ -0,0 +1,173 @@
+
+    val project = getProject(anActionEvent);
+    val branchNameOption = getNameOfBranchUnderAction(anActionEvent);
+    val nonRootBranchOption = branchNameOption.flatMap(bn -> getManagedBranchByName(anActionEvent, bn))
+        .filter(b -> b.isNonRoot())
+        .map(b -> {
+          assert b.isNonRoot() : "branch is root";
+          return b.asNonRoot();
+        });
+    val commitsOption = nonRootBranchOption
+        .map(b -> b.getCommits());
+    val parent = nonRootBranchOption
+        .flatMap(b -> b.getForkPoint()).getOrNull();
+    val syncToParentStatus = nonRootBranchOption.map(b -> b.getSyncToParentStatus()).getOrNull();
+    val gitRepository = getSelectedGitRepository(anActionEvent).getOrNull();
+    val branchName = branchNameOption.getOrNull();
+    val commits = commitsOption.getOrNull();
+
+    if (commits != null && gitRepository != null && parent != null && branchName != null
+        && syncToParentStatus != InSyncButForkPointOff) {
+      val isSquashingCurrent = Option.of(gitRepository.getCurrentBranch()).map(b -> b.getName())
+          .flatMap(cbn -> branchNameOption.map(bn -> bn.equals(cbn))).getOrElse(false);
+      doSquash(project, gitRepository, parent, commits, branchName, isSquashingCurrent);
+    }
+  }
+
+  @UIEffect
+  private void doSquash(Project project,
+      GitRepository gitRepository,
+      ICommitOfManagedBranch parent,
+      List<ICommitOfManagedBranch> commits,
+      String branchName,
+      boolean isSquashingCurrent) {
+    val vcsCommitMetadataAndMessage = commits.foldLeft(
+        new Tuple2<List<VcsCommitMetadata>, String>(List.empty(), ""),
+        (acc, commit) -> new Tuple2<>(
+            acc._1.append(new VcsCommitMetadataAdapter(parent, commit)), "${acc._2}${commit.getFullMessage()}${LS}${LS}"));
+
+    val dialog = new GitNewCommitMessageActionDialog(
+        /* project */ project,
+        /* message */ vcsCommitMetadataAndMessage._2,
+        /* title */ getNonHtmlString("action.GitMachete.BaseSquashAction.dialog.title"),
+        /* dialogLabel */ getNonHtmlString("action.GitMachete.BaseSquashAction.dialog.label"));
+
+    String taskName = isSquashingCurrent
+        ? getString("action.GitMachete.BaseSquashAction.task-title.current")
+        : getString("action.GitMachete.BaseSquashAction.task-title.non-current");
+
+    dialog.show(
+        newMessage -> {
+          new Task.Backgroundable(project, taskName) {
+            @Override
+            @UIThreadUnsafe
+            public void run(ProgressIndicator indicator) {
+              LOG.info("Checking out '${branchName}' branch and squashing it");
+
+              if (!isSquashingCurrent) {
+                CheckoutSelectedBranchAction.doCheckout(project, indicator, branchName, gitRepository);
+              }
+
+              val commitsToSquash = vcsCommitMetadataAndMessage._1.toJavaList();
+              val operationResult = new GitSquashOperation(gitRepository).execute(commitsToSquash, newMessage);
```

Comment:
Out of curiosity... how is the squash performed under the hood? best if you could share the git command line operations

Reply:
`git rebase --interactive --no-autosquash <commit>`

Reply:
😯 okay, if it works, let it be... just out of curiosity, what entry does it leave in `git reflog`?

Reply:
checkout & squash

```
614950c (HEAD -> go-home) HEAD@{0}: rebase (finish): returning to refs/heads/go-home
614950c (HEAD -> go-home) HEAD@{1}: rebase (fixup): ide run
6c327f9 HEAD@{2}: rebase (reword): ide run
6230d66 HEAD@{3}: rebase: fast-forward
85156b4 HEAD@{4}: rebase (start): checkout 85156b4eeb9e22c5d9735908ded91b0f5b2c943b
54d102a (origin/go-home) HEAD@{5}: checkout: moving from make-coffee to go-home
```

Reply:
Okay... indeed what I'd expect after a `git rebase`